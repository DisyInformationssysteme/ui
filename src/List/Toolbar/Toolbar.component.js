import React from 'react';
import { MenuItem, Nav, Navbar, ButtonGroup, DropdownButton } from 'react-bootstrap';
import uuid from 'uuid';
import SelectDisplayMode from './SelectDisplayMode';
import SelectSortBy from './SelectSortBy';
import Pagination from './Pagination';
import Filter from './Filter';
import Action from '../../Actions/Action';
import Icon from '../../Icon';

import theme from './Toolbar.scss';

export function getSubProps(props, component) {
	const subProps = {};
	Object.keys(component.propTypes)
		.filter(key => props[key] !== undefined)
		.forEach((key) => {
			subProps[key] = props[key];
		});
	return subProps;
}

/**
 * @param {object} props react props
 * @example
<Toolbar name="Hello world"></Toolbar>
 */
function Toolbar(props) {
	const displayProps = getSubProps(props, SelectDisplayMode);
	const sortProps = getSubProps(props, SelectSortBy);
	const filterProps = getSubProps(props, Filter);
	const paginationProps = getSubProps(props, Pagination);
	const { id, listActions, onClickAdd } = props;

	const hasProps = p => Object.keys(p).length > 0;
	const hasDisplayMode = hasProps(displayProps);
	const hasSortProps = hasProps(sortProps);
	const hasFilterProps = hasProps(filterProps);
	const hasPaginationProps = hasProps(paginationProps);

	let add;
	if (onClickAdd) {
		add = {
			id: id && `${id}-add-btn`,
			label: 'Add',
			icon: 'fa fa-plus',
			onClick: onClickAdd,
		};
	}
	const dropdownId = id && `${id}-actions-dropdown`;
	return (
		<Navbar componentClass="div" className={theme['tc-list-toolbar']} role="toolbar" fluid>
			{(add || listActions) && (
				<Nav>
					{add && (<Action
						className="navbar-btn"
						bsStyle="success"
						{...add}
					/>)}
					{listActions && (
						<ButtonGroup className="navbar-btn">
							<DropdownButton
								id={dropdownId || uuid.v4()}
								bsStyle="link"
								title="actions"
							>
								{listActions.map((action, index) => {
									const onClick = e => action.onClick(e, action);
									const actionIdentifier = id && action.label.toLowerCase().split(' ').join('-');
									return (
										<MenuItem
											id={id && `${id}-${index}-actions-${actionIdentifier}`}
											key={index}
											onClick={onClick}
											eventKey={index}
										>
											<Icon name={action.icon} />
											{action.label}
										</MenuItem>
									);
								})}
							</DropdownButton>
						</ButtonGroup>
					)}
				</Nav>
			)}
			{hasDisplayMode && (<SelectDisplayMode {...displayProps} />)}
			{hasSortProps && (<SelectSortBy {...sortProps} />)}
			{hasFilterProps && (<Filter {...filterProps} />)}
			{hasPaginationProps && (<Pagination {...paginationProps} />)}
		</Navbar>
	);
}

Toolbar.propTypes = {
	id: React.PropTypes.string,
	onClickAdd: React.PropTypes.func,
	listActions: React.PropTypes.arrayOf(React.PropTypes.object),
	...Filter.propTypes,
	...SelectDisplayMode.propTypes,
	...SelectSortBy.propTypes,
};

export default Toolbar;
