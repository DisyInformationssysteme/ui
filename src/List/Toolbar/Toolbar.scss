// Toolbar styles
$tc-list-toolbar-display-color: $silver !default;
$tc-list-toolbar-background-color: $concrete !default;


.tc-list-toolbar {
	background: $tc-list-toolbar-background-color;

	:global(.navbar-text) {
		color: $tc-list-toolbar-display-color;
		font-weight: 400;
		margin-right: 0;
	}
}
