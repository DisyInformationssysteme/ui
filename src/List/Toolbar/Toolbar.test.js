import React from 'react';
import { shallow } from 'enzyme';

import Toolbar from './Toolbar.component';

const props = {
	onSelectDisplayMode: jest.fn(),
	onSelectSortBy: jest.fn(),
	onFilter: jest.fn(),
	sortOptions: [
		{ id: 'id', name: 'Name' },
		{ id: 'name', name: 'Name' },
	],
	sortBy: 'id',
	sortDesc: true,
};

describe('Toolbar', () => {
	it('should let me click the add btn', () => {
		const tprops = Object.assign({}, props);
		tprops.onClickAdd = jest.fn();
		const wrapper = shallow(
			<Toolbar {...tprops} />
		);
		wrapper.find('.navbar-btn[bsStyle="success"]').simulate('click');
		expect(tprops.onClickAdd).toBeCalled();
	});
});
