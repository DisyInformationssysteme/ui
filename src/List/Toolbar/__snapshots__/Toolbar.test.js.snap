exports[`Toolbar should render 1`] = `
<div
  className="navbar navbar-default"
  role="toolbar">
  <div
    className="container-fluid">
    <ul
      className="nav navbar-nav"
      role={null}>
      <label
        active={undefined}
        activeHref={undefined}
        activeKey={undefined}
        className="navbar-text"
        onSelect={null}>
        Display:
      </label>
      <li
        className="dropdown"
        style={undefined}>
        <a
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle"
          href=""
          id="grid"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button">
          <i
            aria-hidden="true"
            className="fa fa-bars"
            title={null} />
           
          <span
            className="caret" />
        </a>
        <ul
          aria-labelledby="grid"
          className="dropdown-menu"
          role="menu">
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href=""
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              <i
                aria-hidden="true"
                className="fa fa-bars"
                title={null} />
               
              Table
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href=""
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              <i
                aria-hidden="true"
                className="fa fa-list-alt"
                title={null} />
               
              Expanded
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href=""
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              <i
                aria-hidden="true"
                className="fa fa-th-large"
                title={null} />
               
              Tile
            </a>
          </li>
        </ul>
      </li>
    </ul>
    <ul
      className="nav navbar-nav"
      role={null}>
      <label
        active={undefined}
        activeHref={undefined}
        activeKey={undefined}
        className="navbar-text"
        onSelect={null}>
        Sort by:
      </label>
      <li
        className="dropdown"
        style={undefined}>
        <a
          aria-expanded={false}
          aria-haspopup={true}
          className="dropdown-toggle"
          href=""
          id="tc-list-toolbar-sort-by"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button">
          Name
           
          <span
            className="caret" />
        </a>
        <ul
          aria-labelledby="tc-list-toolbar-sort-by"
          className="dropdown-menu"
          role="menu">
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href=""
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              Name
            </a>
          </li>
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href=""
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              Name
            </a>
          </li>
        </ul>
      </li>
      <button
        activeHref={undefined}
        activeKey={undefined}
        className="navbar-btn btn btn-link"
        disabled={false}
        onClick={[Function]}
        onSelect={null}
        role="button"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-sort-desc"
          title={null} />
        DESCENDING
      </button>
    </ul>
    <form
      className="navbar-form navbar-right"
      onSubmit={[Function]}
      role="search">
      <div
        className="form-group">
        <div
          className="input-group">
          <input
            aria-label="Filter"
            className="form-control"
            onChange={[Function]}
            placeholder="Filter"
            type="search" />
        </div>
        <button
          className="btn btn-link"
          disabled={false}
          title="Filter"
          type="submit">
          <i
            aria-hidden="true"
            className="fa fa-search" />
          <span
            className="sr-only">
            Filter
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
